assets/keymaps/vim.json

```diff
@@ -505,5 +505,33 @@
       "enter": "vim::SearchSubmit",
       "escape": "buffer_search::Dismiss"
     }
+  },
+  {
+    // Directory expansion
+    "context": "ProjectPanel && not_editing",
+    "bindings": {
+      "escape": "project_panel::ToggleFocus",
+      "enter": "project_panel::Open",
+      "o": "project_panel::Open",
+      "t": "project_panel::Open",
+      "v": "project_panel::Open",
+      "d": "project_panel::NewDirectory",
+      "%": "project_panel::NewFile",
+      "shift-r": "project_panel::Rename",
+      "m m": "project_panel::Cut",
+      "m c": "project_panel::Copy",
+      "m t": "project_panel::Paste",
```

Comment:
I originally suggested these but, given that they already operate differently than vim, I wonder if it would be less surprising as...

```json
      "shift-d": "project_panel::Delete", # as delete and cut
      "y y": "project_panel::Copy",
      "p": "project_panel::Paste",
```

Thus, the Delete works as both delete and cut. However, it doesn't appear that Delete currently copies the deleted file so it couldn't work without that.

Otherwise, it could also perhaps be...

```json
      "shift-d": "project_panel::Delete",
      "x": "project_panel::Cut",
      "y y": "project_panel::Copy",
      "p": "project_panel::Paste",
```

The `x`, `y y`, and `p` all operating similar as in the Editor. That would require changing the just introduced `x` though. I don't know; just thinking out loud.

Reply:
I think we probably want to keep `x` more like netrw's, but I do like the idea of re-using copy and paste... hmm.

It would be nice to provide shortcuts for yanking the file-path, though it's interesting to note that neither vim nor zed provides these by default (though in vim you can use `yy` or `y$` etc. to yank the filename); and to be honest I'd mostly like to be able to do this from the editor not the project panel.

In the end-game scenario, we want to leave open the path of making `m m` `m c`,  `m t` etc. work correctly (though it's a bizarre UI, some people already know how to use it).

I think we should:

* Continue to support `cmd-c` `cmd-x` `cmd-v` in the project panel for moving files around. 
* Drop the "not quite right" `m X` bindings for now (and keep the issue open if someone wants to add them).
* Add support for `:` in the project panel so you can do `:crp` and `:cpp` for copy relative path and copy path.

I'm on the fence as to whether we should add our own custom vim mode bindings for file management, on the one hand it's nice to have, but on the other it's another bespoke UI that people have to learn to use.

If we were to add vim-style aliases for things, we should pick things that don't conflict with netrw bindings (I'm ok to conflict with editor bindings, because our netrw implementation is not an editor). But it's tricky because netrw has already squatted a bunch of keys (`x`, `p`, `gp`) that would be useful in a cut/copy/paste scenario....

Reply:
That all sounds good. Supporting the things that can be supported cleanly and leaving out the others seems reasonable.

I also wonder if in the far off future it might be possible as an extension to build a _real_ netrw clone that doesn't have to contend with these conflicts, but not sure how far the extension system will go.

At any rate, the project panel is already pretty usable in a netrw-ish way in addition to the other things it can do.

Reply:
It definitely is possible, we'd have to add a few new actions to support things. I think we could do most of them inside the `vim` crate; but first-class vim is a goal for Zed, so we can also make any changes we need to the underlying editor too.


As always, if you'd like to pair on this ever, feel free to book time: https://calendly.com/conradirwin/pairing.